dsl-reference.md

```diff
@@ -1397,7 +1421,7 @@ An event filter is a mechanism used to selectively process or handle events base
 
 | Property | Type | Required | Description |
 |----------|:----:|:--------:|-------------|
-| with | `object` | `yes` | A name/value mapping of the attributes filtered events must define. Supports both regular expressions and runtime expressions.  |
+| with | [`eventProperties`](#event-properties) | `yes` | A name/value mapping of the attributes filtered events must define. Supports both regular expressions and runtime expressions.  |
```

Comment:
I thought I was getting crazy since I couldn't see my previous comments from the other PR. Now I see you reopened it. 

Regarding this property, since we are now defining that it can be regexp, runtime expressions, or `eventProperties`, this should reflect in the schema.

So this `with` should be an `OneOf` there amongst these types.

Reply:
Sorry for the mix-up. When I rebased the branch, GitHub closed my previous PR. I opened a new one and realized afterward that I could have simply reopened the previous one after I readded my commits.

I might be mistaken, but I was under the impression that the statement "Supports both regular expressions and runtime expressions" applies to the properties of `with`, not the object itself. That said, there might be room for discussion about how and where we can use a runtime expression to construct an entire object, but that might be a broader topic than what is addressed in this PR.

@cdavernas @fjtirado @matthias-pichler-warrify, what are your thoughts on this?

Reply:
I would understand this to mean that the individual properties under `with` (such as `id`, `source`, ...) should each support literals, regex and runtime expressions. These are all strings so we would just produce a union of type `string | string | string` which doesn't seem necessary. 
I would keep the `with` object an object. 

regarding regular expressions: How would an implementor determine if a given string is a literal or a regular expression though?

For example if the filter was specified as

```yaml
with:
  type: com.example  
```

if the string was interpreted as a literal only `com.example` would match. As a regex `comaexample` and `com.example.foo` would also match! Are we planning on adding a regex syntax (JSON schema support the ECMA 262 syntax with `format: regex`) or object? Like

```yaml
with:
  type:
    re: ^com\.example

# or

with:
  type: /com\.example/
```

Maybe we should just stick to runtime expressions and use the `test` function: https://jqlang.github.io/jq/manual/#regular-expressions ?

```yaml
with:
  source: ${ test(., "^com\\.example$") }
```

Reply:
That's a good point. It's a very specific pain point considering regular expressions are only supported in `listen` tasks `event filters`.

My understanding was that the value was either a string or a runtime expression. If it's a string, then it's regarded as a regular expression (regexp). But it's not very clear, can introduce undesired matches when not properly escaped, and probably have an bad impact on performance (I would assume string equality to be more performant than regexp, in turn more performant than jq/js expressions.)

For the performance reason I just mentioned (without any proof though, it's only a feeling ^^'), I would rule out relying only on jq/js for processing regexp. Therefore, we need to either consider any string as a regexp or find a way to discriminate the regexp. I must admit I'm not convinced by neither adding a property nor using a specific format but I don't have a better option to suggest. Therefore, I would rather go for the specific format rather than the extra property.

Reply:
> (I would assume string equality to be more performant than regexp, in turn more performant than jq/js expressions.)

My intuition is that the performance difference between string equality and regex matching is negligible compared to `jq` expressions which are much slower.

> Therefore, I would rather go for the specific format rather than the extra property.

If we support regex I would also say that we should use a format such as `/<regex>/`. We would however not be able to rely on the `format: regex` from json schema since this does not have any delimiters